### ISSUE TITLE
Issues for the stage 1 VAE training

### ISSUE THREAD

**Issue by LMozart** [NIRVANALAN/LN3Diff#13]
Dear Authors,

Thank you for your contribution and for open-sourcing this great work!
I recently attempted to reproduce the stage-1 VAE training using the script: `shell_scripts/final_release/train/stage-1-vae/Objaverse/ln3diff-vae-175K_debug.sh` on the pre-processed dataset provided in the README [link](https://entuedu-my.sharepoint.com/:f:/g/personal/yushi001_e_ntu_edu_sg/EtOUTVZNtGxCg3aJnCDkqZcBmH_-OxCGLQBDwqq9ny5Ing?e=BUHiZi). 
To train using all the provided data, I uncommented line 2376 and commented line 2377 in https://github.com/NIRVANALAN/LN3Diff/blob/8599cd341fa13d4fc0848bfe94fe194d73d38fc8/datasets/g_buffer_objaverse.py#L2376-L2377
I trained the VAE on 8 GPUs (40GB each) with a batch size of 1. However, I encountered an issue where the loss became NaN after 850 training steps (sometimes 500 steps). Any ideas or further suggestion why this problem occurs?

P.S., I attached one log file: [log.txt](https://github.com/user-attachments/files/18228817/log.txt)

Best,
Zongrui

**Comment by NIRVANALAN** [comment]
Hi, thanks for your interest in our work. 

1. Overfitting on a single object should be pretty easy, and 1 GPU shall work pretty well. Verifying on the overfitting setting is helpful.
2. Turning of the GAN loss shall help. I enable it in the later stage.
3. You can check our latest work [GaussianAnything](https://github.com/NIRVANALAN/GaussianAnything). I have fixed some key issues there compared to LN3Diff.
4. A VQ-version of LN3Diff will be open sourced at [SAR3D](https://github.com/cyw-3d/SAR3D) soon. The VAE training is pretty similar.

**Comment by LMozart** [comment]
Thank you for your suggestions and prompt reply.
1. Since I’ve included all the objects from the pre-processed dataset you provided as training data, I assume overfitting on a single object won’t occur easily, is that correct?
2. Do you have any suggestions on when I should enable the GAN loss? I’m still trying to understand the code, so if you have already implemented this function, I would appreciate it if you could share an example setup.
3. As I’m new to training VAEs, I’m just curious: is it common for the loss to become NaN during training if the setup is inappropriate?
4. I’ll check out the code of your latest works you mentioned. 
Thank you again for your help.

**Comment by NIRVANALAN** [comment]
1. I mean you can verify the training pipeline by overfitting on a single object first~
2. GAN loss should be enabled only in the later stage. You can disable it by setting ```--trainer_name nv_rec_patch_mvE``` in the bash file. This will enable this trainer: https://github.com/NIRVANALAN/LN3Diff/blob/8599cd341fa13d4fc0848bfe94fe194d73d38fc8/scripts/vit_triplane_train.py#L244
3. If may occur if some training configurations are inappropriate, especially in 3D. Some preliminary debugging on overfitting a single object / a small subset is necessary.

**Comment by LMozart** [comment]
Thank you for your suggestions! They are very helpful. I will try debugging by overfitting on a subset and a single object first. 
In the meantime, just wondering how many steps did you use to train the VAE without GAN loss? And how many steps did you use to train with the GAN loss?

**Comment by NIRVANALAN** [comment]
Hi, usually 60% w/o GAN loss and 40% w/ GAN loss. 500K iterations shall be fine.

**Comment by LMozart** [comment]
Hi. Thank you for your help. I follow your suggestions and conducted several test by overfitting single object and subsets for debugging on one GPU. And I also disable GAN loss by setting trainer as `nv_rec_patch_mvE`. Overfitting on single object seems fine, no NaN values during training. But when I move on to train on a subset, NaN value occurs after 25K steps. I have tested on two subsets (Building and Animals). Any suggestions on this?

P.S. I attached the steps in which the NaN value first occurred for your reference.
[buglog.log](https://github.com/user-attachments/files/18248099/buglog.log)

**Comment by NIRVANALAN** [comment]
That's wired;could you try disabling the LPIPS loss first? Using the animals subset shall be fine to verify this issue. Decreasing the LR and increase the BS also helps.

**Comment by NIRVANALAN** [comment]
Also, are you using FP16 training? Switching to Fp32 or BF16 shall help.

**Comment by LMozart** [comment]
I'm using the default setup and I haven't changed anything except the trainer. I will try again as you recommend. Thanks

**Comment by NIRVANALAN** [comment]
Thanks, and looking forward to your update.

**Comment by LMozart** [comment]
Hi. I’m still working on this issue. Disabling LPIPS didn’t resolve the problem, which aligns with expectations since the first NaN appeared in loss_l2. Enabling use_amp=True helps mitigate the issue to some extent, but NaNs still occur. Interestingly, the training can continue despite this, which is weird. I also found that with AMP enabled, the NaN arises specifically in bg_loss:
https://github.com/NIRVANALAN/LN3Diff/blob/8599cd341fa13d4fc0848bfe94fe194d73d38fc8/nsr/losses/builder.py#L502-L507
Please see the attached log file, where bg_loss:1 indicates there are NaN values.
[buglog.log](https://github.com/user-attachments/files/18265680/buglog.log)

**Comment by NIRVANALAN** [comment]
Hi, thanks for your update. The bg_loss is simply for tuning the reconstruction weight of the background. No idea why it will be NAN. I have not encountered this issue before.

May I ask are you using BF16? The AMP training on V100 with FP16 is pretty unstable from my observation. Using FP16 for the VAE training shall be fine.

**Comment by LMozart** [comment]
Yes, quite strange. I am conducting more experiments to verify this. Previously, I simply set amp=True, which gives me the experiment results I just showed you. I assume the system will automatically using BF16 according to code below: 
https://github.com/NIRVANALAN/LN3Diff/blob/8599cd341fa13d4fc0848bfe94fe194d73d38fc8/nsr/train_util.py#L121
However, I noticed that in some lines you set some operation to FP16. Just to confirm whether I need to change them as BF16. https://github.com/NIRVANALAN/LN3Diff/blob/8599cd341fa13d4fc0848bfe94fe194d73d38fc8/nsr/train_nv_util.py#L1342-L1344

**Comment by LMozart** [comment]
By the way, I found that `(1 - gt_depth_mask).sum() ` sometimes equals 0, which may occur when the patch is fully occupied by the object region, leaving no background region (value 0). By setting a simple if else condition, the problem seems to be solved.

**Comment by NIRVANALAN** [comment]
Interesting, glad to see the problem can be solved by setting a simple if. Dealing with patch rendering is indeed tricky.

**Comment by LMozart** [comment]
Thanks for your help! The training process seems to go well for now. I'll close this issue.
